Engine/DataFeeds/Enumerators/CorporateEventBaseEnumerator.cs

```diff
@@ -0,0 +1,188 @@
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using QuantConnect.Data;
+
+namespace QuantConnect.Lean.Engine.DataFeeds.Enumerators
+{
+    /// <summary>
+    /// Base class for corporate event enumerators. Will call the <see cref="CheckNewEvent"/>
+    /// implementation each time there is a new tradable day. Will merge new events produced with
+    /// underlying enumerator data.
+    /// </summary>
+    public abstract class CorporateEventBaseEnumerator : IEnumerator<BaseData>
+    {
+        private readonly IEnumerator<BaseData> _enumerator;
+        private readonly Queue<BaseData> _auxiliaryData;
+        private bool _underlyingExhausted;
+        private bool _weEmittedData;
+
+        /// <summary>
+        /// The previous data provided by the underlying enumerator
+        /// </summary>
+        protected BaseData PreviousUnderlyingData;
+
+        /// <summary>
+        ///The Subscription data config used by this enumerator
+        /// </summary>
+        protected readonly SubscriptionDataConfig SubscriptionDataConfig;
+
+        /// <summary>
+        /// Creates a new instance
+        /// </summary>
+        /// <param name="enumerator">Underlying enumerator</param>
+        /// <param name="config">The <see cref="SubscriptionDataConfig"/></param>
+        /// <param name="tradableDayNotifier">Tradable dates provider</param>
+        /// <param name="includeAuxiliaryData">True to emit auxiliary data</param>
+        public CorporateEventBaseEnumerator(
+            IEnumerator<BaseData> enumerator,
+            SubscriptionDataConfig config,
+            ITradableDatesNotifier tradableDayNotifier,
+            bool includeAuxiliaryData)
+        {
+            SubscriptionDataConfig = config;
+            _auxiliaryData = new Queue<BaseData>();
+            _enumerator = enumerator;
+            tradableDayNotifier.NewTradableDate += (sender, dateTime) =>
+            {
+                // Call implementation
+                var newEvent = CheckNewEvent(dateTime);
+                if (newEvent != null && includeAuxiliaryData)
+                {
+                    _auxiliaryData.Enqueue(newEvent);
+                }
+            };
+        }
+
+        /// <summary>
+        /// Called each time there is a new tradable day
+        /// </summary>
+        /// <param name="date">The new tradable day value</param>
+        /// <returns>New corporate event, else Null</returns>
+        protected abstract BaseData CheckNewEvent(DateTime date);
```

Comment:
I love the idea of the `IEnumerable<BaseData>`, will do!
What inclined my towards the inheritance was the common dependencies and requirements of the specific corporate events with the base class, such as the `PreviousUnderlyingData` and `SubscriptionDataConfig` properties or the `GetRawClose()`, etc

Reply:
This is an [incredible concept](https://en.wikipedia.org/wiki/Composition_over_inheritance) that every great dev should have a firm grasp on. The desired to use inheritance for code reuse was the hardest anti-pattern for me to break in my own development. These days, I reserve inheritance what I need to _enforce_ certain semantics on a base class, but _never_ for the purpose of code reuse because code reuse can be achieved in a much better way through composition. Consider, instead of an abstract class with multiple base classes, you could have one class (the old abstract class) that takes as a constructor argument an interface (with the old abstract methods), voila, now the code is super reusable because the interface isn't limited to being used by the one base class!

Here's some other good reads on the topic:
* In-depth example of using inheritance and replacing w/ composition by [scottlilly](https://scottlilly.com/c-design-patterns-composition-over-inheritance/)
* Discussion of the [diamon problem] and how composition avoids it by [xiang caohello](https://xiangcaohello.wordpress.com/2012/04/03/composition-over-inheritance-diamond-problem/)
* How composition can help making APIs easier via [stackify](https://stackify.com/oop-concepts-composition/)
* This one has some [funny pictures](https://theburningmonk.com/2015/03/this-is-why-you-need-composition-over-inheritance/) but does a good job showing the worst case scenario when inheritance is taken way too far.

Reply:
Very interesting! I'll re-implement it, using the `composition over inheritance` pattern 👍